geoh5py/shared/concatenation/property_group.py

```diff
@@ -0,0 +1,102 @@
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  geoh5py is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from geoh5py.data import Data
+from geoh5py.groups import PropertyGroup
+
+from .utils import is_concatenated_object
+
+if TYPE_CHECKING:
+    from .object import ConcatenatedObject
+
+
+class ConcatenatedPropertyGroup(PropertyGroup):
+    _parent: ConcatenatedObject
+
+    def __init__(self, parent: ConcatenatedObject, **kwargs):
+        if not is_concatenated_object(parent):
+            raise UserWarning(
+                "Creating a concatenated data must have a parent "
+                "of type Concatenated."
+            )
+
+        super().__init__(parent, **kwargs)
+
+    @property
+    def depth_(self):
+        if self.properties is None or len(self.properties) < 1:
+            return None
+
+        data = self.parent.get_data(  # pylint: disable=no-value-for-parameter
+            self.properties[0]
+        )[0]
+
+        if isinstance(data, Data) and "depth" in data.name.lower():
+            return data
+
+        return None
+
+    @property
+    def from_(self):
+        """Return the data entities defined the 'from' depth intervals."""
+        if self.properties is None or len(self.properties) < 1:
+            return None
+
+        data = self.parent.get_data(  # pylint: disable=no-value-for-parameter
+            self.properties[0]
+        )[0]
+
+        if isinstance(data, Data) and "from" in data.name.lower():
+            return data
+
+        return None
+
+    @property
+    def to_(self):
+        """Return the data entities defined the 'to' depth intervals."""
+        if self.properties is None or len(self.properties) < 2:
+            return None
+
+        data = self.parent.get_data(  # pylint: disable=no-value-for-parameter
+            self.properties[1]
+        )[0]
+
+        if isinstance(data, Data) and "to" in data.name.lower():
+            return data
+
+        return None
+
+    @property
+    def parent(self) -> ConcatenatedObject:
+        return self._parent
+
+    @parent.setter
+    def parent(self, parent):
+        if self._parent is not None:
+            raise AttributeError("Cannot change parent of a property group.")
+
+        if not is_concatenated_object(parent):
+            raise AttributeError(
+                "The 'parent' of a concatenated Data must be of type 'Concatenated'."
+            )
+        parent.add_children([self])
```

Comment:
I think we could avoid circular import and reduce coupling if we let the `parent` do the check for the `isinstance(child, Concatened)` on the `add_children` call, and only do `TYPE_CHECKING` imports for the type hints.

The hierarchy of real imports would be

Concatenated, ConcatenatedPropertyGroup, ConcatenatedData -> ConcatenatedObject -> Concanator

Reply:
Would require to add a `add_children` method on the `ConcatenatedObject` do to the check.

Reply:
See PR #489

Reply:
The problem is not from the parent add_children function but from the object: 
it ensure the parent the user want to set is a concatenated_object.